### PR TITLE
[DRAFT] Add CSV storage

### LINKED ISSUE #532: [CSVStorage] Implement storage that supports CSV
You have a separate schema file, which will probably be a file with the SQL extension.
ex) `{table_name}.schema.sql`
We read it in and create a table.

Then, when scan data, schema validation is performed. (Wouldn't it be enough to verify only on the first run?)

### PR DESCRIPTION
**Pull request by MRGRAVITY817**
## Resolves #532 
This new storage implementation will have the following functionalities:
- [ ] Read CSV file from the path like reading a table.
- [ ] Check if the CSV file has an interpretable schema using `{table_name}.schema.sql` file.
- [ ] Mutate the original CSV file, or make a copy of mutated result.
- [ ] [...will add more]

## WIP 🚧

### CODE REVIEW

File: storages/csv-storage/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "gluesql_csv_storage"
```
Comment (ever0de) [inline storages/csv-storage/Cargo.toml:2]: It would be better to use -.

```suggestion
name = "gluesql-csv-storage"
```

**Review by ever0de — COMMENTED** [review]
😀

File: storages/csv-storage/Cargo.toml (hunk 1 above)
Comment (MRGRAVITY817) [reply storages/csv-storage/Cargo.toml:2]:
```toml
[package]
name = "gluesql_memory_storage"
```
Hmm... but other storage subcrates in this project use snake case.

Comment (ever0de) [reply storages/csv-storage/Cargo.toml:2]: memory storage has been deployed before so it is :cry:

Comment (ever0de) [reply storages/csv-storage/Cargo.toml:2]: ref: <https://github.com/gluesql/gluesql/blob/main/storages/shared-memory-storage/Cargo.toml#L2>

Comment (panarch) [reply storages/csv-storage/Cargo.toml:2]: Oh, you don't need to care about underscore cases. You can use `-` dash.
I published `gluesql_memory_storage` and `gluesql_sled_storage` before we change our project structure into workspace-based.
And.. `crates.io` does not distinguish `_` and `-` so I wasn't able to rename `gluesql_memory_storage` to `gluesql-memory-storage`.


image

**Comment by MRGRAVITY817** [comment]
@panarch 

Initial structure that I've been thinking for `schema.toml` 

```toml
[table.users]
path = "example/data/users.csv"
columns.id = "Int128"
columns.name = "Text"
columns.age = "Uint8"
columns.role = "Text"

[table.orders]
path = "example/data/orders.csv"
columns.id = "Int128"
columns.name = "Text"
columns.orderer_id = "Int128"
columns.restaurant_id = "Int128"
columns.quantity = "Uint8"
columns.price = "Float"
columns.status = "Text"

```

**Comment by MRGRAVITY817** [comment]
Or maybe this one is better, at least more easy to parse using `toml` crate.

```toml
[[tables]]
name = "users"
path = "example/data/users.csv"
columns = [
	{ name = "id",   type = "Int128", options = ["PKey"]    },
	{ name = "name", type = "Text",   options = ["NotNull"] },
	{ name = "age",  type = "Unit8"                         },
	{ name = "role", type = "Text",   options = ["NotNull"] },
]
```

**Pull request closed without merging**